Fix super error mishandling uncurried function

JonoPrest opened the pull request:
Closes #6662 

Super errors were not handling uncurried functions properly.

```res
let apply = (fn: (. unit) => option<int>) => fn(. ())

let _ = apply(Some(1))
```

outputted:

```
The constructor Some does not belong to type function$
  
  This variant expression is expected to have type . unit => option<int>
```

whereas it should have outputted

```
  This value might need to be wrapped in a function that takes an extra
  parameter of type unit
                                                                                                                                            
  Here's the original error message
  This has type: option<'a>
  But it's expected to have type: (. unit) => option<int>
```

There were a number of place where uncurried functions were not being handled here and I suspect there's a number of gremlins still around.

JonoPrest commented on jscomp/build_tests/super_errors/expected/UncurriedArgsNotApplied.res.expected:
```diff
@@ -0,0 +1,15 @@
+  [1;31m3[0m [2m│[0m let _ = apply([1;31mSome(1)[0m)
+  4 [2m│[0m 
+
+  This value might need to be [1;33mwrapped in a function that takes an extra
+  parameter[0m of type unit
+  
+  [1;33mHere's the original error message[0m
+  This has type: [1;31moption<'a>[0m
+  But it's expected to have type: [1;33m(. unit) => option<int>[0m
```
This now matches the super error that was outputted when a curried function was defined

JonoPrest commented on jscomp/ml/ast_uncurried.ml:
```diff
@@ -69,14 +69,20 @@ let coreTypeIsUncurriedFun (typ : Parsetree.core_type) =
-let typeIsUncurriedFun = Ast_uncurried_utils.typeIsUncurriedFun
-
-let typeExtractUncurriedFun (typ : Parsetree.core_type) =
+let coreTypeExtractUncurriedFun (typ : Parsetree.core_type) =
```
Renamed from typeExtractUncurriedFun coreTypeExtractUncurriedFun to follow naming convention in this file. I I have redefined  typeExtractUncurriedFun below. (coreTypeIsUncurriedFun is named above fore the same paramater type, and for functions with Types.type_expr as a param use typeIsUncurredFun and typeExtractUncurriedFun respectively)

JonoPrest commented on jscomp/ml/error_message_utils.ml:
```diff
@@ -204,7 +204,8 @@ let typeClashContextMaybeOption ty_expected ty_res =
       {Types.desc = Tconstr (typePath, _, _)} )
     when Path.same Predef.path_option typePath
-         && Path.same expectedPath Predef.path_option = false ->
+         && Path.same expectedPath Predef.path_option = false 
+         && Path.same expectedPath Predef.path_uncurried = false ->
```
Here was another gremlin I came across where you get a suggestion to unwrap your option. For the example in the tests/descriptioon you get:

```
  This value might need to be wrapped in a function that takes an extra
  parameter of type unit
                                                                                                                                            
  Here's the original error message
  This has type: option<'a>
  But it's expected to have type: (. unit) => option<int>
                                                                                                                                            
  Possible solutions:
  - Unwrap the option to its underlying value using `yourValue->Belt.Option.getWithDefault(someDefaultValue)`
```

Where the possible solution is not meant to be for functions but uncurried functions still pattern match on this case.

JonoPrest commented on jscomp/ml/typecore.ml:
```diff
@@ -307,7 +307,9 @@ let extract_concrete_record env ty =
 let extract_concrete_variant env ty =
   match extract_concrete_typedecl env ty with
-    (p0, p, {type_kind=Type_variant cstrs}) -> (p0, p, cstrs)
+    (p0, p, {type_kind=Type_variant cstrs}) 
+      when not (Ast_uncurried.typeIsUncurriedFun ty)  
+      -> (p0, p, cstrs)
```
This was the first issue, leading to the initial bad super error being displayed. The uncurried function type would extract as a concrete variant since it matched this case when it should not.

mununki replied:
I'm sure you've read and know more about this, but instead of throwing an exception with Not_found if uncurried, what about an implementation that extracts the uncurried case? Something like this.
```ml
if Ast_uncurried_utils.typeIsUncurriedFun ty then
   extract_concrete_variant env (Ast_uncurried.typeExtractUncurriedFun ty)
else (p0, p, cstrs)
```

JonoPrest replied:
Hey @mununki thanks for the comment! Yes I think that feel a lot nicer except that in practice it would make no difference to this function since if it is an uncurried function the extracted type would not be Type_variant? So happy to add the refactor to make it more future proof but otherwise it should never hit a case where it is an uncurried function and it can extract a concrete variant

mununki replied:
Sorry for my misunderstanding, if so why do we need that when clause by then?

JonoPrest replied:
> Sorry for my misunderstanding, if so why do we need that when clause by then?

It's because an uncurried function is represented by this pattern

```ml
Tconstr (Pident {name = "function$"}, [tArg; _], _)
```

where tArg is the underlying type of `Tarrow`

So when trying to extract the concrete type as a variant, it matches the case of Tconstr, but is not actually variant it's essentially just a wrapper around the underlying function. So we get a false positive.

JonoPrest commented on jscomp/ml/typecore.ml:
```diff
@@ -662,6 +664,9 @@ let rec collect_missing_arguments env type1 type2 = match type1 with
       | Some res -> Some ((label, argtype) :: res)
       | None -> None
     end
+  | t when Ast_uncurried_utils.typeIsUncurriedFun t -> 
+    let typ = Ast_uncurried.typeExtractUncurriedFun t in 
+    collect_missing_arguments env typ type2
```
This was the second issue where it should have given the suggestion:

```
  This value might need to be wrapped in a function that takes an extra
  parameter of type unit
                                                                                                                                            
  Here's the original error message
  This has type: option<'a>
  But it's expected to have type: (. unit) => option<int>
                                                                                                                                            
```
it only outputted

```
  This has type: option<'a>
  But it's expected to have type: (. unit) => option<int>
```

since it would not match on the cases which treat it as  a function. We first need to extract the type of the uncurried function.

cristianoc reviewed (commented):
Thanks for the fix. This looks great!

Looks like something went off when updating the changelog.

JonoPrest commented:
> Thanks for the fix. This looks great!
> 
> Looks like something went off when updating the changelog.

Thanks @cristianoc, I think there was an issue when I rebased on GH and then added another commit locally. I've fixed up the commit rebased and pushed.

zth commented:
@JonoPrest impressive work, well done! 😄
